v3io/dataplane/transport/httpclient.py

```diff
@@ -47,37 +51,55 @@ def __init__(self, logger, endpoint=None, max_connections=None, timeout=None, ve
         self._get_status_and_headers = self._get_status_and_headers_py3
 
     def close(self):
-        # Ignore redundant calls to close
-        if not self._free_connections:
-            return
-
-        connections = []
-        while not self._free_connections.empty():
-            conn = self._free_connections.get()
-            connections.append(conn)
-        # In case anyone tries to reuse this object, we want them to get an error and not hang
-        self._free_connections = None
-        self._logger.debug(f"Closing all {len(connections)} v3io transport connections")
-        for conn in connections:
-            conn.close()
+        with self._lock:
+            # Avoid redundant calls to close
+            if self._closed:
+                return
+
+            # Mark as closed before draining the queue to prevent race conditions
+            self._closed = True
+
+            connections = []
+            with contextlib.suppress(queue.Empty):
+                while not self._free_connections.empty():
+                    conn = self._free_connections.get_nowait()
+                    connections.append(conn)
+            self._logger.debug(f"Closing all {len(connections)} v3io transport connections")
+            for conn in connections:
+                try:
+                    conn.close()
+                except Exception as e:
+                    self._logger.debug(f"Error closing connection: {e}")
 
     def requires_access_key(self):
         return True
 
     def send_request(self, request):
-        if not self._free_connections:
-            raise RuntimeError("Cannot send request on a closed client")
+        with self._lock:
+            if self._closed:
```

Comment:
Consider using double-checked locking, especially here since it's on the hot path.

Reply:
Double-checked locking is always only to avoid locking unnecessarily, unrelated to the content of the block.